### ISSUE TITLE
A list of objects automatically add an item even if all of the fields are empty

### ISSUE THREAD

**Issue by alexgmin** [bhch/django-jsonform#38]
```
        schema={
            'type': 'array',
            'items': {
                'type': 'dict',
                'keys': {
                    'foo': {
                        'type': 'string',
                    },
                    'bar': {
                        'type': 'string',
                    },
                },
            },
        },
```

if you edit an instance that has [] in database, even if you don't touch any of the fields, it automatically stores
```
[{"foo": "", "bar": ""}]
```

While I understand that's how django manages empty string fields, setting them to an empty string instead of null, it doesn't make sense for it to happen in cases like this. It's causing us problems because then our frontend/api tries to return this stuff and it shows an empty item.

I'm not sure how to handle this case, but here's three suggestions:
- If min_items is 0, don't show an empty item by default, but if a user clicks on add item and doesn't add anything the issue still happens, which is not really the expected behavior.
- A new option for dicts to only be stored if any of the values is not null (nested objects are a problem)
- Let the user handle it by extending the field

**Comment by bhch** [comment]
Okay, if the field is initially empty and is unedited, I will add some checks to prevent adding empty items.

> but if a user clicks on add item and doesn't add anything the issue still happens, which is not really the expected behavior.

You can't really know if the user intends to keep an empty item or not. So, if the user manually adds an empty item, that should be kept in the data.

> A new option for dicts to only be stored if any of the values is not null (nested objects are a problem)

I'll provide some validation hooks for this. Before saving the data, your custom validation functions will be called where you can remove non-empty dict keys.

I will try and include this in the next release (maybe next week).

**Comment by alexgmin** [comment]
> You can't really know if the user intends to keep an empty item or not. So, if the user manually adds an empty item, that should be kept in the data.

That's fair, it's a weird case were both storing as empty or not storing could be interpreted as valid.

> I will try and include this in the next release (maybe next week).

Thank you so much.

**Comment by bhch** [comment]
@alexgmin Hi, I'm working on this right now. I'm looking for some suggestions. If the array is unedited, should I save an empty array (`[ ]`) in the db or keep the field completely empty (`null`)?

**Comment by alexgmin** [comment]
I think `[]` makes more sense too, but what if the field has a default? Shouldn't it be that?

**Comment by bhch** [comment]
Ah, thanks for reminding me that.

**Comment by alexgmin** [comment]
So the only change you're making is making it empty in min items isn't set, not setting it to the default if it's empty?

What you suggest is feasible, although I'm not a fan of overriding the save method for this type of things, since the save method isn't called in bulk updates. What I've been currently doing for this type of custom things, like https://github.com/bhch/django-jsonform/issues/18 is creating a custom field and overriding the get_prep_value or the pre_save methods

```
class ActionTimestampsJSONFieldWithSchema(JSONFieldWithSchema):
    # Remove values that are empty strings
    def pre_save(self, model_instance, add):
        value = super().pre_save(model_instance, add

        for timestamp_key in value.copy().keys():
            if value[timestamp_key] == '':
                del value[timestamp_key]

        return value
```

I also have a similar case in which I want an object to be null if none of boolean keys inside are true, because that makes querying that field much simpler.

If the min items issue is resolved I can work with the rest.

**Comment by bhch** [comment]
> So the only change you're making is making it empty in min items isn't set, not setting it to the default if it's empty?

Yes, if a default is provided it will use that value instead of being blank.

I'll also add a `pre_save_hook` parameter for your use case. If you have a better suggestion, do let me know.

I'll try to release the new version by tomorrow.

**Comment by alexgmin** [comment]
A `pre_save_hook`would allow me to remove all of of my overrided fields, so I would be extremely grateful.

**Comment by bhch** [comment]
I've released [v2.10.0](https://github.com/bhch/django-jsonform/releases/tag/v2.10.0) which fixes this.

Please upgrade and test out the new version.

**Comment by alexgmin** [comment]
@bhch Perfect, will do. I also sent you an email about a private issue, could you check it?
